File: validate_html.py
import re

with open('E:/YPWI ABSENSI/public/admin-dashboard.html', 'r') as f:
    content = f.read()

# Count opening and closing tags
opens = len(re.findall(r'<(?!\!--|!DOCTYPE|html)', content))
closes = len(re.findall(r'</', content))
print(f'Opening tags: {opens}, Closing tags: {closes}')

# Check for duplicate IDs
ids = re.findall(r'id="([^"]+)"', content)
dups = [x for x in ids if ids.count(x) > 1]
if dups:
    print(f'Duplicate IDs: {set(dups)}')
else:
    print('No duplicate IDs')

# Check script tag balance
script_opens = content.count('<script')
script_closes = content.count('</script>')
print(f'Script tags: {script_opens} open, {script_closes} close')

# Check for missing </div>
opens_div = content.count('<div')
closes_div = content.count('</div>')
print(f'Div tags: {opens_div} open, {closes_div} close')

# Check form tags
opens_form = content.count('<form')
closes_form = content.count('</form>')
print(f'Form tags: {opens_form} open, {closes_form} close')

# Check for img tags without closing
self_closing = len(re.findall(r'<img[^>]*\s*/>', content))
print(f'Self-closing img tags: {self_closing}')

File: validate_html.js
const fs = require('fs');
const content = fs.readFileSync('E:/YPWI ABSENSI/public/admin-dashboard.html', 'utf8');

// Check for duplicate IDs
const idRegex = /id="([^"]+)"/g;
const ids = [];
let match;
while ((match = idRegex.exec(content)) !== null) {
  ids.push(match[1]);
}
const idCounts = {};
ids.forEach(function(id) { idCounts[id] = (idCounts[id] || 0) + 1; });
const dups = Object.entries(idCounts).filter(function(entry) { return entry[1] > 1; });
if (dups.length > 0) {
  console.log('Duplicate IDs:');
  dups.forEach(function(entry) { console.log('  ' + entry[0] + ': ' + entry[1] + ' times'); });
} else {
  console.log('No duplicate IDs found');
}

// Count tags
const openDivs = (content.match(/<div/g) || []).length;
const closeDivs = (content.match(/<\/div>/g) || []).length;
console.log('Div open: ' + openDivs + ', close: ' + closeDivs);

const openScripts = (content.match(/<script/g) || []).length;
const closeScripts = (content.match(/<\/script>/g) || []).length;
console.log('Script open: ' + openScripts + ', close: ' + closeScripts);

const openForms = (content.match(/<form/g) || []).length;
const closeForms = (content.match(/<\/form>/g) || []).length;
console.log('Form open: ' + openForms + ', close: ' + closeForms);

// Check for img without closing
const imgTags = (content.match(/<img/g) || []).length;
console.log('Img tags: ' + imgTags);

// Check unclosed tags
const liOpens = (content.match(/<li/g) || []).length;
const liCloses = (content.match(/<\/li>/g) || []).length;
console.log('Li open: ' + liOpens + ', close: ' + liCloses);